### PR TITLE
LG-2715: Correctly allow redirect_uris to app-specific uris

### PR DESCRIPTION
**Pull request by solipet**
**Why?** to correctly allow `redirect_uris` that provide deep links into native applications.

Previously, we declared a uri invalid unless it had a domain, while app-specific uris have only a scheme and path.

### CODE REVIEW

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -14,3 +20,24 @@
     c.syntax = :expect
   end
 end
+
+class TestServiceProvider
```
Comment (zachmargolis) [inline spec/spec_helper.rb:24]: can thing go in its own file? I would never expect to find a class defined in `spec_helper` like this?

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -14,3 +20,24 @@
...
+
+  def initialize(**args)
+    super
+  end
+
+  def new_record?
+    true
+  end
+
+  def self._reflect_on_association(foo)
```
Comment (zachmargolis) [inline spec/spec_helper.rb:40]: I would call the argument `_` if it's unused

File: spec/identity/service_provider_validation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      saml_client_cert: nil_cert
+    )
+  end
+
+  it 'validates a valid service provider' do
+    allow_any_instance_of(ActiveRecord::Validations::UniquenessValidator).to receive(:validate_each).and_return(true)
+
+    service_provider.valid?
+    expect(service_provider).to be_valid, "SP not valid due to #{service_provider.errors.messages}"
+  end
```
Comment (zachmargolis) [inline spec/identity/service_provider_validation_spec.rb:59]: could we add a few more specs for examples of valid and invalid `uri_valid?` values? to exercise/document the branches in the code

File: spec/spec_helper.rb (hunk 1 above)
Comment (solipet) [reply spec/spec_helper.rb]: Yup!

File: spec/spec_helper.rb (hunk 2 above)
Comment (solipet) [reply spec/spec_helper.rb]: good call

File: lib/identity_validations/service_provider_validation.rb [hunk 4]
```diff
@@ -63,7 +64,8 @@ def saml_client_cert_is_x509_if_present
 
     def uri_valid?(uri)
       parsed_uri = URI.parse(uri)
-      parsed_uri.scheme.present? && parsed_uri.host.present?
+      return false unless parsed_uri.scheme.present?
+      /https?/ =~ parsed_uri.scheme ? parsed_uri.host.present? : true
```
Comment (zachmargolis) [inline lib/identity_validations/service_provider_validation.rb:68]: another quick thought --- there's the `native` attribute in the IDP database for ServiceProvider, what if we made sure that non-HTTP urls have `native` set to true?

File: spec/identity/service_provider_validation_spec.rb (hunk 3 above)
Comment (solipet) [reply spec/identity/service_provider_validation_spec.rb]: Yup!

File: lib/identity_validations/service_provider_validation.rb (hunk 4 above)
Comment (solipet) [reply lib/identity_validations/service_provider_validation.rb:68]: That would be fine to add as a new validation, but I don't know that it is consistently applied. We could make a story to add it?

Comment (zachmargolis) [reply lib/identity_validations/service_provider_validation.rb:68]: ya for sure, if you want to track that as a follow-up that's fine

**Pull request merged**